File: src/domino/simplify.cc
#include <simplify.h>

namespace domino {
std::vector<ExprSimplifyPattern> ExprSimplifier::patterns_;

bool ExprSimplifyMatchPattern(Expr target, Expr pattern) {
  ExprSimplifyPatternMatcher matcher;
  return matcher(target, pattern);
}

std::unordered_map<Var, Expr> GetExprSimplifyMatchPatterns(Expr target, Expr pattern) {
  ExprSimplifyPatternMatcher matcher;
  bool match = matcher(target, pattern);
  if (match) {
    return matcher.getMapping();
  } else {
    return std::unordered_map<Var, Expr>();
  }
}

Expr SubstituteExpr(Expr expr, std::unordered_map<Var, Expr> mapping) {
  ExprSubstituter suber(mapping);
  return suber(expr);
}

Stmt SubstituteStmt(Stmt stmt, std::unordered_map<Var, Expr> mapping) {
  auto suber = std::make_shared<ExprSubstituter>(mapping);
  StmtMutator mutator(suber);
  return mutator(stmt);
}

Block SubstituteBlock(Block block, std::unordered_map<Var, Expr> mapping) {
  auto suber = std::make_shared<ExprSubstituter>(mapping);
  auto stmt_mutator = std::make_shared<StmtMutator>(suber);
  BlockMutator block_mutator(stmt_mutator);
  return block_mutator(block);
}

Arch SubstituteArch(Arch arch, std::unordered_map<Var, Expr> mapping) {
  auto suber = std::make_shared<ExprSubstituter>(mapping);
  auto stmt_mutator = std::make_shared<StmtMutator>(suber);
  auto block_mutator = std::make_shared<BlockMutator>(stmt_mutator);
  ArchMutator arch_mutator(block_mutator);
  return arch_mutator(arch);
}

IRBase SubstituteIR(IRBase ir, std::unordered_map<Var, Expr> mapping) {
  if (ir.as<ExprNode>().defined()) {
    return SubstituteExpr(ir.as<ExprNode>(), mapping);
  } else if (ir.as<StmtNode>().defined()) {
    return SubstituteStmt(ir.as<StmtNode>(), mapping);
  } else if (ir.as<BlockNode>().defined()) {
    return SubstituteBlock(ir.as<BlockNode>(), mapping);
  } else if (ir.as<ArchNode>().defined()) {
    return SubstituteArch(ir.as<ArchNode>(), mapping);
  } else {
    throw std::runtime_error("The IR type is not supported in substitution.");
  }
}

Expr SimplifyExpr(Expr expr) {
  ExprSimplifier sim;
  return sim.Visit(expr);
}

IRBase Simplify(IRBase ir) {
  auto sim = std::make_shared<ExprSimplifier>();
  IRMutator mutator(sim);
  return mutator.Visit(ir);
}

}  // namespace domino

File: python/domino/program_ir/simplify.py
from dominoc import ir
from .scalar_expr import _to_expr
from .stmt import _to_stmt
from .block import _to_block

__all__ = ["expr_simplify_match_pattern",
           "get_expr_simplify_match_patterns",
           "substitute_expr", "simplify_expr", "substitute_ir", "simplify"]

expr_simplify_match_pattern = ir.expr_simplify_match_pattern
get_expr_simplify_match_patterns = ir.get_expr_simplify_match_patterns


def substitute_expr(expr, mapping):
    mapping = {k: _to_expr(v) for k, v in mapping.items()}
    expr = _to_expr(expr)
    return ir.substitute_expr(expr, mapping)


def substitute_stmt(ir_tree, mapping):
    mapping = {k: _to_expr(v) for k, v in mapping.items()}
    ir_tree = _to_stmt(ir_tree)
    return ir.substitute_stmt(ir_tree, mapping)


def substitute_block(ir_tree, mapping):
    mapping = {k: _to_expr(v) for k, v in mapping.items()}
    ir_tree = _to_block(ir_tree)
    return ir.substitute_block(ir_tree, mapping)


def substitute_ir(ir_tree, mapping):
    mapping = {k: _to_expr(v) for k, v in mapping.items()}
    return ir.substitute_ir(ir_tree, mapping)


def simplify_expr(expr):
    return ir.simplify_expr(expr)


def simplify(irm):
    return ir.simplify(irm)
